lib/datadog/statsd/message_buffer.rb

```diff
@@ -92,6 +123,33 @@ def ensure_sendable!(message_size)
       def bytesize_threshold
         @bytesize_threshold ||= (max_payload_size - PAYLOAD_SIZE_TOLERANCE * max_payload_size).to_i
       end
+
+      def create_flush_thread
+        flush_thread = Thread.new(&method(:flush_loop))
+        flush_thread.name = 'Statsd MessageBuffer' if flush_thread.respond_to?(:name=)
+        flush_thread
+      end
+
+      def flush_loop
+        last_flush_time = current_time
+        until @closed
+          @mon.synchronize do
+            @cv.wait(flush_interval - (current_time - last_flush_time))
+            last_flush_time = current_time
+            flush
```

Comment:
`flush` also uses `@mon.synchronize`, so I'd expect this to deadlock (unless Monitor is re-entrant, but I don't see that in its API docs).

Reply:
Yes, this is why I use `Monitor` instead of `Mutex`. `Monitor` is re-entrant.
I couldn't find the official document written in English, but [the document written in Japanese](https://docs.ruby-lang.org/ja/3.1/library/monitor.html) states that it provides the feature allowing the same thread to lock multiple times:

> monitor は以下のような Thread::Mutex としての機能も提供します。
> 
> * lock の持ち主がスレッドである Mutex / 何度も lock できる Mutex
>    - lock したスレッドを Mutex 側が覚えていて
>    - そのスレッドがもう一度 lock しようとしてもブロックしない
>    - synchronize は lock していなかったら通常どおり、 自分が lock していたら ただ yield するだけ (lock/unlockもしない)
>    - unlock はそのスレッドだけができる

Reply:
You can try the behavior with the following script:

```ruby
require 'monitor'

mon = Monitor.new
mon.synchronize do
  mon.synchronize do
    puts 'hello'
  end
end
```

Reply:
OK, that's a bit bizarre but I'm glad it works :)